app/src/services/permit.ts

```diff
@@ -173,22 +134,79 @@ const service = {
   },
 
   /**
-   * @function updatePermit
-   * Updates a Permit
+   * @function upsertPermit
+   * Upsert a Permit
    * @param {Permit} data Permit object
    * @returns {Promise<Permit | null>} The result of running the update operation
    */
-  updatePermit: async (data: Permit) => {
+  upsertPermit: async (data: Permit) => {
     try {
-      const response = await prisma.permit.update({
-        include: {
-          permit_type: true
-        },
-        data: { ...permit.toPrismaModel(data), updated_by: data.updatedBy },
-        where: {
-          permit_id: data.permitId
+      const permitDataWithId = { ...data, permitId: data.permitId || uuidv4() };
+
+      // Upsert the permit
+      const response = await prisma.$transaction(async (trx) => {
+        const response = await trx.permit.upsert({
+          include: {
+            permit_type: true
+          },
+          where: {
+            permit_id: permitDataWithId.permitId
+          },
+          update: { ...permit.toPrismaModel(permitDataWithId), updated_by: permitDataWithId.updatedBy },
+          create: {
+            ...permit.toPrismaModel({ ...permitDataWithId }),
+            created_by: permitDataWithId.createdBy,
+            updated_by: permitDataWithId.updatedBy
+          }
+        });
+
+        // Delete any permit tracking that is not in the new data
+        await trx.permit_tracking.deleteMany({
+          where: {
+            permitId: permitDataWithId.permitId,
+            permitTrackingId: {
+              notIn: permitDataWithId.permitTracking
+                ?.map((x: PermitTracking) => x.permitTrackingId)
+                .filter((x) => x) as number[]
```

Comment:
If possible, we should try to avoid using Typescript as assertions, as they should be considered tools of last resort. You should be able to cast this earlier in the chain as a number like this, which properly asserts the return type of map and cascades further down:

```ts
?.map<number>((x: PermitTracking) => x.permitTrackingId)
.filter((x) => x)
```

Reply:
In that case, you need to do one of the following:
- Declare that x.permitTrackingId is not going to be a null value with `x.permitTrackingId!` with trailing exclamation
- `?.map<number | undefined>((x: PermitTracking) => x.permitTrackingId)`